core/src/id.rs

```diff
@@ -0,0 +1,126 @@
+    }
+
+    fn next() -> u64 {
+        NEXT_ID.fetch_add(1, atomic::Ordering::Relaxed)
+    }
+
+    /// Creates a unique [`Id`].
+    ///
+    /// This function produces a different [`Id`] every time it is called.
+    pub fn unique() -> Self {
+        let id = Self::next();
+
+        Self(Internal::Unique(id))
+    }
+}
+
+// Not meant to be used directly
+#[cfg(feature = "a11y")]
+impl From<u64> for Id {
+    fn from(value: u64) -> Self {
+        Self(Internal::Unique(value))
+    }
+}
+
+// Not meant to be used directly
+#[cfg(feature = "a11y")]
+impl Into<NonZeroU128> for Id {
+    fn into(self) -> NonZeroU128 {
+        match &self.0 {
+            Internal::Unique(id) => NonZeroU128::try_from(*id as u128).unwrap(),
+            Internal::Custom(id, _) => {
+                NonZeroU128::try_from(*id as u128).unwrap()
+            }
+            // this is a set id, which is not a valid id and will not ever be converted to a NonZeroU128
+            // so we panic
+            Internal::Set(_) => {
+                panic!("Cannot convert a set id to a NonZeroU128")
+            }
+        }
+    }
+}
+
+impl ToString for Id {
+    fn to_string(&self) -> String {
+        match &self.0 {
+            Internal::Unique(_) => "Undefined".to_string(),
+            Internal::Custom(_, id) => id.to_string(),
+            Internal::Set(_) => "Set".to_string(),
+        }
+    }
+}
+
+// XXX WIndow IDs are made unique by adding u64::MAX to them
+/// get window node id that won't conflict with other node ids for the duration of the program
+pub fn window_node_id() -> NonZeroU128 {
+    std::num::NonZeroU128::try_from(
+        u64::MAX as u128
+            + NEXT_WINDOW_ID.fetch_add(1, atomic::Ordering::Relaxed) as u128,
+    )
+    .unwrap()
+}
+
+// TODO refactor to make panic impossible?
+#[derive(Debug, Clone, Eq, Hash)]
+/// Internal representation of an [`Id`].
+pub enum Internal {
+    /// a unique id
+    Unique(u64),
+    /// a custom id, which is equal to any [`Id`] with a matching number or string
+    Custom(u64, borrow::Cow<'static, str>),
+    /// XXX Do not use this as an id for an accessibility node, it will panic!
+    /// XXX Only meant to be used for widgets that have multiple accessibility nodes, each with a
+    /// unique or custom id
+    /// an Id Set, which is equal to any [`Id`] with a matching number or string
+    Set(Vec<Self>),
+}
+
+impl PartialEq for Internal {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Unique(l0), Self::Unique(r0)) => l0 == r0,
+            (Self::Custom(l0, l1), Self::Custom(r0, r1)) => {
+                l0 == r0 || l1 == r1
+            }
+            // allow custom ids to be equal to unique ids
+            (Self::Unique(l0), Self::Custom(r0, _))
+            | (Self::Custom(l0, _), Self::Unique(r0)) => l0 == r0,
+            (Self::Set(l0), Self::Set(r0)) => l0.iter().zip(r0).any(|(l, r)| l == r),
```

Comment:
This is comparing vec elements pairwise, so it should be equivalent to just testing equality of the vectors, but ignoring additional elements if one vec is longer than the other? Probably not what's wanted here.

Just dropping this case should work, I think. It will match the `(Self::Set(l0), r) | (r, Self::Set(l0))` case, and match it again when it recurses.

Reply:
Or if we want to test that they are the same, in the same order, `l0 == r0` works.

Or for set equality ignoring order... I guess use a `HashSet` or `BTreeSet` instead of a `Vec`.

Reply:
Ya I can do `l0 == r0` for now instead. I think that maybe I called it a set when it should really just be a list. If it turns out that order doesn't matter, I'll change it to a HashSet or BTreeSet.